docs/decex/fraud_proofs.md

```diff
@@ -211,103 +226,116 @@ There are several variants of why `inboxed_bundles` in the receipt can be wrong:
+- `maybe_domain_runtime_code_proof`: for the runtime code if it is not still present in the state.
+- `bundle_with_proof`: including `bundle_index` index of mismatched bundle; `bundle` bundle body; and `bundle_storage_proof` storage proof.
 
 **Verifier checks:**
 
 1. Verify the `bad_receipt` with `bad_receipt_hash` exists.
-2. Obtain from the host function the `bundle_body`  of the bundle in question.
-3. Compute `valid_bundle_digest` of the `bundle_body`.
-4. Compare that the digest is different from the one in `bad_receipt` ⇒ Accept the fraud proof and punish the producer of `bad_receipt`.
-5. If the digest is same ⇒ Ignore the fraud proof.
+2. Verify `mmr_proof` and obtain the corresponding state root for the consensus block number in the receipt.
+3. Get the domain runtime code that used to derive the target receipt: if the runtime code is still present in the state then get it from the state, otherwise from the `maybe_domain_runtime_code_proof` storage proof with MMR proof.
+4. Verify the `bundle_storage_proof`.
+5. Obtain from `bundle_with_proof` the `bundle` body of the bundle in question.
+6. Compute `valid_bundle_digest` from extrinsics in the `bundle` using domain runtime code.
+7. Compare that the digest is different from the one in `bad_receipt`at the given `bundle_index` ⇒ Accept the fraud proof and punish the producer of `bad_receipt`.
+8. If the digest is same ⇒ Ignore the fraud proof.
 
 ### True/False Invalid Bundle
 
 **Prover provides:**
 
 - `domain_id`: ID of the domain this fraud proof targets.
 - `bad_receipt_hash`: the targeted invalid ER.
+- `mmr_proof`: MMR proof for the parent consensus block of the receipt.
+- `maybe_domain_runtime_code_proof`: for the runtime code if it is not still present in the state.
 - `bundle_index`: index of mismatched bundle
 - `invalid_bundle_type(extrinsic_index)`: the `InvalidBundleType` with the mismatched extrinsic
-- `extrinsic_inclusion_proof`: proof-of-inclusion of the extrinsic in the `invalid_bundle_type`
 - `is_true_invalid_fraud_proof`: whether the variant of this proof is `TrueInvalid`(`true`) or `FalseInvalid`(`false`)
+- `proof_data`: specific data for the fraud proof variant.
 
 **Verifier checks:**
 
 1. Verify the `bad_receipt` with `bad_receipt_hash` exists.
-2. Determine the fraud proof variant. If `is_true_invalid_fraud_proof==true` then it’s `TrueInvalid`, else `FalseInvalid`.
-3. Determine which scenario the fraud proof corresponds to (if any) based on its structure:
+2. Verify `mmr_proof` and obtain the corresponding state root for the consensus block number in the receipt.
+3. Get the domain runtime code that used to derive the target receipt: if the runtime code is still present in the state then get it from the state, otherwise from the `maybe_domain_runtime_code_proof` storage proof with MMR proof.
+4. Determine the fraud proof variant. If `is_true_invalid_fraud_proof==true` then it’s `TrueInvalid`, else `FalseInvalid`.
```

Comment:
I remember we still need to handle the case where extrinsic index is out of range. This is currently not in I believe and there is an issue that talk about how to handle this.

Reply:
It's this one https://github.com/subspace/subspace/issues/2287

Reply:
I see a decision has been made. Should we include this in the iteration of the spec or do you prefer we update the spec later ?

Reply:
Let's keep it separate, this PR already has too much stuff